### ISSUE TITLE
CS0019: Operator '!=' cannot be applied to operands of type 'TEnum' and 'TEnum' where TEnum: Enum

### ISSUE THREAD

**Issue by voroninp** [dotnet/csharplang#2025]
```CSharp
private static object[] GetEnumLookupData<TEnum>() where TEnum: Enum
{
    var data = Enum.GetValues(typeof(TEnum))
        .OfType<TEnum>()
        .Where(e => e != default(TEnum))
        .Select(e => new {Id = e, Name = e.ToString()})
        .ToArray();

    return data;
}
```

This does not compile, but returns a error:
> CS0019: Operator '!=' cannot be applied to operands of type 'TEnum' and 'TEnum'

What am I missing here? Isn't it clear that `TEnum` is a value type?

Sure, I can use `EqualityComparer<TEnum>.Default.Equals(...)`, but still ...

**Comment by YairHalberstadt** [comment]
Why not just `!e.Equals(default (TEnum))`

**Comment by chrisoverzero** [comment]
>What am I missing here? Isn't it clear that TEnum is a value type?

No, because it doesn’t say so in the function definition. What changes if you change the definition to this?

```csharp
private static object[] GetEnumLookupData<TEnum>() where TEnum: struct, Enum
```

**Comment by voroninp** [comment]
@chrisoverzero This does not help. 
@YairHalberstadt  Why do I need boxing one more time? `EqualitComparer` exists just for this.

**Comment by HaloFour** [comment]
The C# compiler doesn't treat the `System.Enum` constraint in any special manner.  It doesn't even consider it, by itself, to be a value type, because `System.Enum` itself is not a value type and you may use `System.Enum` as the generic argument.  If you combine the constraints `System.Enum` and `struct` then you do force the compiler and runtime to only allow actual enum instances, but the compiler still doesn't consider them to be any different than normal structs and doesn't permit arithmetic operations.  The support for `System.Enum` as a generic type constraint in C# 7.x was only to lift the imposed limitation and allow what the CLR already permitted.

**Comment by chrisoverzero** [comment]
>This does not help.

Sure it did. Now you have a different problem, one caused by EF. Not all `enum` values can be cast to `Int32`.

**Comment by voroninp** [comment]
@chrisoverzero  No it did not. Try the first version without `int` yourself.

**Comment by chrisoverzero** [comment]
@voroninp You are playing multiple solutions at once, confusing both me and the issue.

To get `default(TEnum)` to do what you want it to do – that is, to be the 0 value of the `enum` represented by `TEnum’ – the type signature I suggested will work. This solves these problems you came up against:

>Isn't it clear that TEnum is a value type?

>[…] but `default` generates porbably just a `null` […]

This frees you up to apply one of the equality checking suggestions from this thread, the instance method or the default equality comparer.

How to solve the EF problem, I have no idea. An `enum` as a primary key strikes me as super odd, though. Perhaps you could use the name of the value as the key, but this is outside the scope of csharplang.

**Comment by voroninp** [comment]
@chrisoverzero > the type signature I suggested will work

Did you try it?! It does not compile.

```CSharp
private static object[] GetEnumLookupData<TEnum>() where TEnum: struct, Enum
{
    var data = Enum.GetValues(typeof(TEnum))
        .OfType<TEnum>()
        .Where(e => e != default(TEnum))
        .Select(e => new {Id = e, Name = e.ToString()})
        .ToArray();

    return data;
}
```

**Comment by chrisoverzero** [comment]
@voroninp I addressed the problem of equality separately.
